src/scaffolding.rs

```diff
@@ -34,7 +35,7 @@ ffi_support::define_bytebuffer_destructor!({{ ci.ffi_bytebuffer_free().name() }}
 // of items as declared in the rust code, but no harm will come from it.
 
 {% for e in ci.iter_enum_definitions() %}
-    unsafe impl uniffi::support::ViaFfi for {{ e.name() }} {
+    unsafe impl uniffi::support::ViaFfi for {{ e.name()|decl_name_rs }} {
```

Comment:
Rather than doing this conversion when generating the Rust scaffolding, what do you think about the following approach:

* Decree that the naming convention in the IDL should be the same as the naming convention in the Rust code (but allow users to write it differently if they want, maybe with a warning in future).
* When parsing the IDL, canonicalize all names to the rust convention as part of building the `ComponentInterface` struct (this would be the place where we warn about it in future).
* Allow each foreign language binding to choose whether and how it wants to deviate from that convention to fit in with the norms of that language.

Alternately, we could just insist that the names in Rust and IDL must match exactly, however you've written them, and not try to do any conversion.

Keeping the IDL and the Rust more tightly bound makes sense to me in a vague kind of way, especially if we imagine that one day we might general definitions using inline rust macros rather than a separate IDL file. It would also mean there are no "control knobs" for callers to fiddle with when interfacing between the IDL and the Rust code, while allowing the possibility of such knobs when generating the bindings (which there is also a placeholder `Config` class for in the kotlin and python bindings generators).

Thoughts?

Reply:
If I understand correctly, I agree with your: 

> insist that the names in Rust and IDL must match
> while allowing the possibility of such knobs when generating the bindings

There will be some back pressure from the hand written rust from the linter or rustc so that the rust and WebIDL will naturally tend to conform to a rust naming convention— because the rust compiler will insist, rather than `uniffi`.  If users really want to have funky rust identifiers, they'll have to add their own annotations in rust.

The foreign language bindings are a view onto the WebIDL, so I think they should have a little bit of leeway to convert/mangle names to suite the language/project. I don't think we should do it without configuration in the long run because:

 * collisions are still possible, (the webIDL could define `foo_bar` and `fooBar`, which would both convert to the same methods). This might correspond to a normalize + check step before generating any bindings.
 * generating source-backward-compatible bindings for existing projects may still be an aim for the project.

If all this is desirable, then we should also consider doing the same for the property names.

Reply:
> collisions are still possible, (the webIDL could define foo_bar and fooBar, which would both convert to the same methods).
> This might correspond to a normalize + check step before generating any bindings.

This is a good point; I think I'd be interested in trying out a normalize+check step when processing the IDL. In the same way that we'd error out if you tried to define two methods with exactly the same name, we could error out if you try to define two methods that would be ambiguous under common case conversions. (And it seems easier to error out now and loosen the rules later, than to start with loose rules and try to tighten them up later).

Reply:
Filed #13